libs/data/bbhnet/data/dataloader.py

```diff
@@ -0,0 +1,372 @@
+        """Sample a batch of kernels from the background data
+
+        Randomly sample kernels from the interferometer
+        background timeseries in a uniform manner. Removing
+        the `independent` kwarg for now to leverage the
+        more general sample_kernels function.
+
+        TODO: figure out how best to generalize sample_kernels
+            to make use of pre-sampled idx
+        """
+
+        hanford_kernels = sample_kernels(
+            self.hanford_background, self.kernel_size, self.batch_size
+        )
+        livingston_kernels = sample_kernels(
+            self.livingston_background, self.kernel_size, self.batch_size
+        )
+
+        # interweave these kernels along the 0th axis so that
+        # a reshape puts them in the right channel dimension
+        kernels = zip(hanford_kernels, livingston_kernels)
+        kernels = [i for j in kernels for i in j]
+        kernels = torch.stack(kernels, dim=0)
+        return kernels.reshape(self.batch_size, 2, -1)
+
+    def whiten(self, X: torch.Tensor) -> torch.Tensor:
+        """Detrend and time-domain filter an array
+        Use the time-domain filters built from the
+        background data used to initialize the dataset
+        to whiten an array of data.
+        """
+
+        # do a constant detrend along the time axis,
+        # transposing to ensure that the last two dimensions
+        # of the original and dimension-reduced tensors match.
+        # TODO: will using X.mean(axis=-1, keepdims=True)
+        # allow us to avoid these transposes?
+        X = X.transpose(2, 0)
+        X = X - X.mean(axis=0)
+        X = X.transpose(0, 2)
+        X *= self.whitening_window
+
+        # convolve the detrended data with the time-domain
+        # filters constructed during initialization from
+        # the background data, using groups to ensure that
+        # the convolution is performed independently for
+        # each interferometer channel
+        X = torch.nn.functional.conv1d(
+            X, self.whitening_filter, groups=2, padding="same"
+        )
+
+        # scale by sqrt(2 / sample_rate) for some inscrutable
+        # signal processing reason beyond my understanding
+        return X * (2 / self.sample_rate) ** 0.5
+
+    def __iter__(self):
+        self._batch_idx = 0
+        return self
+
+    def __next__(self):
+        if self._batch_idx >= self.batches_per_epoch:
+            raise StopIteration
+
+        # create an array of all background
+        X = self.sample_from_background()
+
+        # create a target tensor, marking all the glitch data as 0.
+        y = torch.zeros((self.batch_size,))
+
+        # replace some of this data with glitches if
+        # we have glitch data to use
+        if self.glitch_sampler is not None:
+            hanford_glitches, livingston_glitches = self.glitch_sampler.sample(
+                self.num_glitches, self.kernel_size
+            )
+

```

Comment:
To clarify, this portion is creating "incoherent" glitches. i.e., no data-sample has a glitch in both Hanford and Livingston. 

In the future, might be worth considering adding some fraction of samples w/ glitches in both Hanford and Livingston.